### PR TITLE
test_template_on_error: rewrite test cases in criterion

### LINKED ISSUE #2578: convert ut test_template_on_error to criterion
Use Criterion to rewrite this unit test.

### PR DESCRIPTION
**Pull request by mehul-m-prajapati**
fixes #2578

### CODE REVIEW

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
do nothing -> CI won't start)

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
do nothing -> CI won't start)

**Comment by Kokan** [comment]
@kira-syslogng ok to test

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/template/tests/test_template_on_error.c [hunk 1]
```diff
@@ -25,17 +25,17 @@
 #include <libgen.h>
 
 #include "template/templates.h"
-#include "testutils.h"
+#include <criterion/criterion.h>
 #include "apphook.h"
 
 #define assert_on_error_parse(on_error,expected)                        \
```
Comment (furiel) [inline lib/template/tests/test_template_on_error.c]: You can eliminate these macros due to criterion:
For now we a single test with five asserts, four for success case, one for failure. This single test should be split into two tests: a parametrized criterion test for the four success cases, and one regular test for the failure case.

About parameterezed test, I mean those that are defined with `ParameterizedTest` and `ParameterizedTestParameters`.

Comment (mehul-m-prajapati) [reply lib/template/tests/test_template_on_error.c]: Hi @furiel ,

I am exploring ParameterizedTest and will make changes as you suggested.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/template/tests/test_template_on_error.c (hunk 1 above)
Comment (furiel) [reply lib/template/tests/test_template_on_error.c]: Thank you! The test looks beautiful. :)

**Pull request merged**